src/Components/Web.JS/src/Platform/WebAssemblyResourceLoader.ts

```diff
@@ -0,0 +1,195 @@
+    );
+    return { name, url, response, data };
+  }
+
+  logToConsole() {
+    const cacheLoadsEntries = Object.values(this.cacheLoads);
+    const networkLoadsEntries = Object.values(this.networkLoads);
+    const cacheResponseBytes = countTotalBytes(cacheLoadsEntries);
+    const networkResponseBytes = countTotalBytes(networkLoadsEntries);
+    const totalResponseBytes = cacheResponseBytes + networkResponseBytes;
+    const linkerDisabledWarning = this.bootConfig.linkerEnabled ? '%c' : '\n%cThis application was built with linking (tree shaking) disabled. Published applications will be significantly smaller.';
+
+    console.groupCollapsed(`%cblazor%c Loaded ${toDataSizeString(totalResponseBytes)} resources${linkerDisabledWarning}`, 'background: purple; color: white; padding: 1px 3px; border-radius: 3px;', 'font-weight: bold;', 'font-weight: normal;');
+
+    if (cacheLoadsEntries.length) {
+      console.groupCollapsed(`Loaded ${toDataSizeString(cacheResponseBytes)} resources from cache`);
+      console.table(this.cacheLoads);
+      console.groupEnd();
+    }
+
+    if (networkLoadsEntries.length) {
+      console.groupCollapsed(`Loaded ${toDataSizeString(networkResponseBytes)} resources from network`);
+      console.table(this.networkLoads);
+      console.groupEnd();
+    }
+
+    console.groupEnd();
+  }
+
+  async purgeUnusedCacheEntriesAsync() {
+    // We want to keep the cache small because, even though the browser will evict entries if it
+    // gets too big, we don't want to be considered problematic by the end user viewing storage stats
+    const cachedRequests = await this.cache.keys();
+    const deletionPromises = cachedRequests.map(async cachedRequest => {
+      if (!(cachedRequest.url in this.usedCacheKeys)) {
+        await this.cache.delete(cachedRequest);
+      }
+    });
+
+    return Promise.all(deletionPromises);
+  }
+
+  private async tryValidateAndCacheAsync(name: string, cacheKey: string, expectedContentHash: string, networkResponse: Response): Promise<ArrayBuffer> {
+    const responseBuffer = await networkResponse.arrayBuffer();
+
+    // Now is an ideal moment to capture the performance stats for the request, since it
+    // only just completed and is most likely to still be in the buffer. However this is
+    // only done on a 'best effort' basis. Even if we do receive an entry, some of its
+    // properties may be blanked out if it was a CORS request.
+    const performanceEntry = getPerformanceEntry(networkResponse.url);
+    const responseBytes = (performanceEntry && performanceEntry.encodedBodySize) || undefined;
+    this.networkLoads[name] = { responseBytes };
+
+    // crypto.subtle is only enabled on localhost and HTTPS origins
+    // We only write to the cache if we can validate the content hashes
+    if (typeof crypto !== 'undefined' && !!crypto.subtle) {
+      await assertContentHashMatchesAsync(name, responseBuffer, expectedContentHash);
```

Comment:
Why is the hash check important? How do you get to a state where the hash could is no longer valid and how would a user fix it?

Reply:
It’s important because caching the wrong version of a file is disastrous - you can’t recover from that without manually clearing caches, which is impossible to explain to most end users. The hash check protects users from this situation, so at worst they can reload the page to recover. 

Consider visiting the site at the same moment that a new version is being deployed. Half the assemblies are the old versions, then the second half comes from the new versions. This likely means a broken app, or at least undefined behaviour. Without the hash check this combination of files would get cached and that user would be stuck with the errors. With the hash check, the initial load fails which is good as it prevents the undefined behaviour, then on the next visit the user gets and caches a working state. 

Same general situation if you have multiple front end servers and no client affinity. If you’re incrementally rolling out updates, we want to detect the case where you load an inconsistent set of resources because the requests were handled by an incompatible collection of servers.

Reply:
BTW we only check hashes before caching under a key derived from that hash. Once a file is in the cache we know its a correct match and we don’t check it again.